chain/epoch_manager/src/lib.rs

```diff
@@ -1496,22 +1491,20 @@ impl EpochManager {
         let mut cur_hash = *last_block_hash;
         let mut overwrite = false;
         while cur_hash != aggregator.last_block_hash || epoch_change {
-            // Avoid cloning
-            let prev_hash = *self.get_block_info(&cur_hash)?.prev_hash();
-            let prev_height = self.get_block_info(&prev_hash).map(|info| *info.height());
+            let cur_block_info = self.get_block_info(&cur_hash)?;
+            let prev_hash = cur_block_info.prev_hash();
 
-            let block_info = self.get_block_info(&cur_hash)?;
-            if block_info.epoch_id() != epoch_id || block_info.prev_hash() == &CryptoHash::default()
-            {
+            if cur_block_info.epoch_id() != epoch_id || prev_hash == &CryptoHash::default() {
                 // This means that we reached the previous epoch and still hasn't seen
                 // `aggregator.last_block_hash` and therefore implies either a fork has happened
                 // or we are at the start of an epoch. In this case, the new aggregator should
                 // overwrite the old one.
                 overwrite = true;
                 break;
             }
-            new_aggregator.update(block_info, &epoch_info, prev_height?);
-            cur_hash = *block_info.prev_hash();
+            let prev_height = *self.get_block_info(prev_hash)?.height();
+            new_aggregator.update(&cur_block_info, &epoch_info, prev_height);
+            cur_hash = *prev_hash;
```

Comment:
This loop can be further optimised so that this second get_block_info is not necessary. What’s previous block in one iteration becomes current block in the next iteration so really this loop should be something along the lines of:

```rust
let mut current_block = get_block_info(last_block_hash);
loop {
  ...
  let prev_block = get_block_info(current_block.prev_hash());
  ...
  current_block = prev_block;
}
```

It’s best if we wait for #6531 before going with this PR I think.  It
does a bit of refactoring around this method.

Reply:
Yeah, I've noticed that, and decided to leave it for the follow up, for fear of actually messing up the logic.